url458e40ff41_mmacfadden_java_baldwin_ga_tgz-pJ8-edu_sdsu_cs652_si_baldwin_SimulationEngineJ8/src/edu/sdsu/cs652/si/baldwin/RankProportionalSelector.java

```diff
@@ -0,0 +1,260 @@
+	 * removed from the wheel so that it cannot be selected again.
+	 * 
+	 * @param a_generator
+	 *            the random number generator to be used during the spinning
+	 *            process
+	 * @param a_fitnessValues
+	 *            an array of fitness values of the respective Chromosomes
+	 * @param a_counterValues
+	 *            an array of total counter values of the respective Chromosomes
+	 * @param a_chromosomes
+	 *            the respective Chromosome instances from which selection is to
+	 *            occur
+	 * @return selected Chromosome from the roulette wheel
+	 * 
+	 * @author Neil Rotstan
+	 * @author Klaus Meffert
+	 * @since 1.0
+	 */
+	private IChromosome selectChromosome(RandomGenerator generator) {
+		int selectedSlot = generator.nextInt(this.totlaSlots);
+		IChromosome result = null;
+		int currentSlot = 0;
+		for (int i = 0; i < this.slotCounts.size() && result == null; i++) {
+			if (selectedSlot <= currentSlot) {
+				result = this.chromosomes.get(i);
+				this.chromosomes.remove(i);
+				this.totlaSlots = this.totlaSlots - this.slotCounts.remove(i);
+				return result;
+			} else {
+				currentSlot += this.slotCounts.get(i);
+			}
+		}
+		return this.chromosomes.get(this.chromosomes.size() - 1);
+	}
+
+	/**
+	 * Empty out the working pool of Chromosomes.
+	 * 
+	 * @author Neil Rotstan
+	 * @since 1.0
+	 */
+	public synchronized void empty() {
+		this.chromosomes.clear();
+		this.slotCounts.clear();
+	}
+
+	/**
+	 * @return always false as some Chromosome's could be returnd multiple times
+	 * 
+	 * @author Klaus Meffert
+	 * @since 2.0
+	 */
+	public boolean returnsUniqueChromosomes() {
+		return true;
+	}
+
+	/**
+	 * Not supported by this selector! Please do not use it.
+	 * 
+	 * @param a_doublettesAllowed
+	 *            do not use
+	 * 
+	 * @author Klaus Meffert
+	 * @since 2.0
+	 */
+	public void setDoubletteChromosomesAllowed(final boolean a_doublettesAllowed) {
+		throw new IllegalStateException("Weighted roulette selector does not"
+				+ " support this parameter," + " please do not use it!");
+	}
+
+	/**
+	 * @return TRUE: doublette chromosomes allowed to be added by the selector
+	 * 
+	 * @author Klaus Meffert
+	 * @since 2.0
+	 */
+	public boolean getDoubletteChromosomesAllowed() {
+		return true;
+	}
+
+	/**
+	 * @return deep clone of this instance
+	 * 
+	 * @author Klaus Meffert
+	 * @since 3.2
+	 */
+	public Object clone() {
+		try {
+			RankProportionalSelector result = new RankProportionalSelector(
+					getConfiguration());
+			result.chromosomes = new ArrayList<IChromosome>(this.chromosomes);
+			result.slotCounts = new ArrayList<Integer>(this.slotCounts);
+			return result;
+		} catch (InvalidConfigurationException iex) {
+			throw new CloneException(iex);
+		}
+	}
+
+	public boolean equals(Object o) {
+		RankProportionalSelector other = (RankProportionalSelector) o;
+		if (other == null) {
+			return false;
+		}
+		
+		return (this.chromosomes.equals(other.chromosomes)) &&
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **257**. At this line the following field is not protected by synchronization: `chromosomes`. Your code typically protects this field with synchronization on object: `RankProportionalSelector.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `RankProportionalSelector.this`. For example, like at lines: **68** and **195**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **68**, **106**, **177**, **178**, **185**, **185**, and **195**.